### DISCUSSION TITLE
Passing two parameters to a function, one of which is multipart: Multipart

### DISCUSSION THREAD

**Discussion by Lucky-49** [tokio-rs/axum#2726] [General]
Hello everyone!
I have a problem: I need to convert the file sent by the client. I tried to write like this:

```
let route_convert = Router::new()
        .route("/upload", post(upload_file));

async fn upload_file(
      Path(output_format): Path<&str>,
      multipart: Multipart,
    ) -> impl IntoResponse {
```

but the compiler throws an error. How do I correctly accept a file from the client and an additional parameter that will indicate which format to convert to?

**Comment by mladedav** [comment]
You might want to post the error you're getting.

**Comment by jplatte** [comment]
Extractors can't borrow anything. Use `Path<String>`, not `Path<&str>`.

**Reply by Lucky-49** [reply]
How easily the problem was solved!!!! Thank you very much!!

**Comment by Lucky-49** [comment]
It wasn't that simple. (( 
Now I get an error and don't understand what I'm doing wrong. 

```
error[E0277]: the trait bound `fn(axum::extract::Path<std::string::String>, Multipart) -> impl Future<Output = std::result::Result<DownloadFile, error::Error>> {handler_convert_file}: Handler<_, _>` is not satisfied
   --> src\main.rs:22:47
    |
22  |         .route("/upload/:output_format", post(handler_convert_file));
    |                                          ---- ^^^^^^^^^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for fn item `fn(Path<String>, Multipart) -> impl Future<Output = ...> {handler_convert_file}`

```

Please help me figure it out.
The code is here [https://github.com/Lucky-49/untitled123](url)

**Reply by mladedav** [reply]
Try adding [`#[debug_handler]`](https://docs.rs/axum-macros/latest/axum_macros/attr.debug_handler.html) to the handler function. It should provide a better error message than `rustc`.

Also, your link is broken, it points somewhere else but even if I copy it it returns 404.

**Reply by Lucky-49** [reply]
https://github.com/Lucky-49/untitled123

**Reply by mladedav** [reply]
Still 404. Is it possible the repository is private?

Did you try adding the `debug_handler` attribute to your function?

**Reply by Lucky-49** [reply]
Sorry, I didn't pay attention to the privacy of the repository.
Yes, I used `#[debug_handler]`.

` ) -> Result<DownloadFile> {
   |          ^^^^^^ the trait `IntoResponse` is not implemented for `DownloadFile`, which is required by `std::result::Result<DownloadFile, error::Error>: IntoResponse`

 ) -> Result<DownloadFile> {
   |          ^^^^^^ required by this bound in `check`

 .route("/upload/:output_format", post(handler_convert_file));
    |                                          ---- ^^^^^^^^^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for fn item `fn(Path<String>, Multipart) -> impl Future<Output = ...> {handler_convert_file}`

`

**Comment by wangziling** [comment]
Hi Vitaliy~ After pulling your demo codes.

I found that maybe you missing to impl `IntoResponse` for your `DownloadFile` struct in `src/web/routes_files.rs :12`.

```rust
impl IntoResponse for DownloadFile {
    fn into_response(self) -> axum::response::Response {
        use axum::http::HeaderValue;

        let mut res = self.file_data.0.into_response();
        res.headers_mut().insert(
            "content-disposition",
            // File name.
            HeaderValue::from_bytes(self.file_name.as_bytes()).unwrap(),
        );

        res
    }
}
```

And your `handler_convert_file` method has two ways to fix the errors.
1. Remove the `Ok()` wrapper.'Cause the compiler didn't know What the `Error` is when you are using the `std::result::Result<T, Error>::Ok()`, you didn't point it out.
![image](https://github.com/tokio-rs/axum/assets/24818903/91d03201-9357-47d4-8eeb-5427a50cc0aa)

2. Using the `Result` what is made alias by yourself.
```rust
Result::Ok(build_response_file)
```
![image](https://github.com/tokio-rs/axum/assets/24818903/b3b489cc-df35-464e-863b-74866918c146)

Or just modify the method's return type.
![image](https://github.com/tokio-rs/axum/assets/24818903/e5e2eba3-d45e-4316-a89b-8893c00638f5)

And then you need to `impl IntoResponse` trait for customized `Error` enum.
![image](https://github.com/tokio-rs/axum/assets/24818903/a79584de-7ae8-49b1-8b87-77e605550dcb)

**Reply by wangziling** [reply]
If you want to make sure to catch all the errors in `handler_convert_file` method.

Suggests that you'd better to remove all the maybe-panic `.unwrap()` invokings, just using the `?` operation.
And then you need enhance the `Error` handling events.
[https://github.com/rust-lang/crates.io/blob/main/src/util/errors.rs](https://github.com/rust-lang/crates.io/blob/main/src/util/errors.rs) I'm following the `crates.io`'s error handling events.

**Reply by Lucky-49** [reply]
Thank you very much for your detailed answer!!!  You helped solve my problem!
